# Week 1 · Topic 3 — Numbers and Arithmetic

**Box we are building: a first look at PROCESS**

This is where a program stops storing and starts working things out.

> **Already know this?** Skip to the *Topic 3 Challenge*.

---
## 1. The operators

Run this and compare all seven at once.

In [1]:
a = 17
b = 5

print("a + b  =", a + b)
print("a - b  =", a - b)
print("a * b  =", a * b)
print("a / b  =", a / b)
print("a // b =", a // b)
print("a % b  =", a % b)
print("a ** b =", a ** b)

a + b  = 22
a - b  = 12
a * b  = 85
a / b  = 3.4
a // b = 3
a % b  = 2
a ** b = 1419857


### Your turn
Set `a = 23` and `b = 4` and predict all seven answers **before** running it.

In [ ]:
a = 23
b = 4 
print("a + b  =", a + b) #27
print("a - b  =", a - b) #19
print("a * b  =", a * b) #92
print("a / b  =", a / b) #5.75
print("a // b =", a // b) #5
print("a % b  =", a % b) #3
print("a ** b =", a ** b) #279841




a + b  = 27
a - b  = 19
a * b  = 92
a / b  = 5.75
a // b = 5
a % b  = 3
a ** b = 279841


---
## 2. `/` always gives a float

Even when it divides exactly.

In [3]:
print(10 / 2)
print(type(10 / 2))

5.0
<class 'float'>


`//` throws away the fraction. `%` gives what is left over.

In [4]:
print(10 // 3)
print(10 % 3)

3
1


Together they split a number into parts. This is more useful than it looks.

In [5]:
total_minutes = 137

hours = total_minutes // 60
minutes = total_minutes % 60

print(total_minutes, "minutes is", hours, "hours and", minutes, "minutes")

137 minutes is 2 hours and 17 minutes


### Your turn
500 seconds is how many whole minutes, and how many seconds left over?

In [6]:
total_seconds = 500 

minutes = total_seconds // 60
seconds = total_seconds % 60

print(total_seconds, "seconds is", minutes, "minutes and", seconds, "seconds")


500 seconds is 8 minutes and 20 seconds


One split is useful. Chaining two splits, from the biggest unit down, is what you actually need for a stopwatch or a duration.

In [7]:
total_seconds = 4831

hours = total_seconds // 3600
remaining = total_seconds % 3600

minutes = remaining // 60
seconds = remaining % 60

print(total_seconds, "seconds =", hours, "hours,", minutes, "minutes,", seconds, "seconds")

4831 seconds = 1 hours, 20 minutes, 31 seconds


Same idea twice: split off the biggest unit first with `// 3600` and `% 3600`, then split what is *left over* into minutes and seconds the same way you already did above.

### Your turn
9137 seconds is how many hours, minutes and seconds? Do it in two chained steps, the way the cell above does.

In [8]:
total_seconds = 9137
hours = total_seconds // 3600
remaining = total_seconds % 3600

minutes = remaining // 60
seconds = remaining % 60

print(total_seconds, "seconds =", hours, "hours,", minutes, "minutes,", seconds, "seconds")


9137 seconds = 2 hours, 32 minutes, 17 seconds


---
## 3. Order of operations

Python follows normal maths rules: `**` first, then `*` `/` `//` `%`, then `+` `-`.

In [9]:
print(2 + 3 * 4)
print((2 + 3) * 4)

14
20


> **Habit worth forming:** use brackets even when you do not need them. They cost nothing and remove all doubt.

### Your turn
The next cell gives the wrong answer. The intention was *(10 plus 6) divided by 2*. Fix it.

In [11]:
print(10 + 6 / 2)

print ((10 + 6) / 2)


13.0
8.0


---
## 4. The calculation you will write most often

*What share of the whole is this?*

In [15]:
value = 23.7
limit = 20

difference = value - limit
percent = (value / limit) * 100

print("Difference:", difference)
print("Percent of limit:", percent)

Difference: 3.6999999999999993
Percent of limit: 118.5


### Your turn
A server has 87 GB used out of 120 GB.
Calculate and print the percentage used, and how many GB are free.

In [17]:
total_storage = 120 
used_storage = 87
free_storage = total_storage - used_storage
percent_used = (used_storage / total_storage) * 100

print("Free storage is :", free_storage)
print("Percent of storage used is: ", percent_used ,"%")


Free storage is : 33
Percent of storage used is:  72.5 %


---
## 5. Shorthand

Changing a variable using its own value is so common it has a short form.

In [18]:
count = 0

count = count + 1     # the long way
print(count)

count += 1            # exactly the same thing
print(count)

count += 5
print(count)

1
2
7


### Your turn
Start `total` at 100. Subtract 30, then double it, using the short forms. Print the result.

In [19]:
total = 100 
total -= 30 
total *= 2

print(total)

140


---
## Topic 3 Challenge

A batch of 250 records was checked. 38 of them failed.

Calculate and print, each on its own line:

- how many passed
- the percentage that failed
- the percentage that passed

Use well-named variables. Do not type any answer you could calculate.

In [20]:
total_records = 250
failed  = 38
passed = total_records - failed
print("Number of passed records:", passed)
print("percebtage that failed :" , (failed / total_records) * 100)
print("percebtage that passed :" , (passed / total_records) * 100)



Number of passed records: 212
percebtage that failed : 15.2
percebtage that passed : 84.8


---
## Same idea, three fields

Almost every calculation you write in year one is *"what share of the whole is this?"*

| Field | Question |
|---|---|
| **AI / Data Science** | what percentage of rows are missing a value |
| **Cyber Security** | what percentage of login attempts failed |
| **IT** | what percentage of disk space is used |

All three are `(part / total) * 100`.


---
## Common mistakes

| Mistake | What happens | Fix |
|---|---|---|
| `2 + 3 * 4` when you meant `(2 + 3) * 4` | wrong answer, **no error** | use brackets |
| dividing by zero | `ZeroDivisionError` — program stops | check the total is not 0 |
| expecting `10 / 2` to be `5` | it is `5.0` | `/` always gives a float |
| `3.6999999999999993` | not a bug | format it: `f"{x:.2f}"` |

## Go deeper

`examples/01_operators.py` — all seven operators, `//` and `%`, precedence
`examples/02_calculations.py` — the percentage pattern in all three fields


---
## Recap

| You can now | Using |
|---|---|
| do arithmetic | `+ - * / // % **` |
| split a number into parts | `//` and `%` |
| chain splits into hours, minutes, seconds | `// 3600` / `% 3600`, then `// 60` / `% 60` on what's left |
| control the order | brackets |
| find a percentage | `(part / total) * 100` |
| update a variable | `+=` |

**Next:** Topic 4 — getting values from the user and reporting back.